lib/cli/src/commands/package/download.rs

```diff
@@ -93,43 +90,53 @@ impl PackageDownload {
+            PackageSource::Ident(PackageIdent::Named(id)) => {
                 let endpoint = self.env.registry_endpoint()?;
-                let version = version.to_string();
+                let version = id.version_or_default().to_string();
                 let version = if version == "*" { None } else { Some(version) };
-
-                (
-                    full_name,
-                    version,
-                    endpoint,
-                    self.env.get_token_opt().map(|x| x.to_string()),
+                let full_name = id.full_name();
+                let token = self.env.get_token_opt().map(|x| x.to_string());
+
+                let package = wasmer_registry::query_package_from_registry(
+                    endpoint.as_str(),
+                    &full_name,
+                    version.as_deref(),
+                    token.as_deref(),
                 )
-            }
-            PackageSpecifier::Url(url) => {
-                bail!("cannot download a package from a URL: '{}'", url);
-            }
-            PackageSpecifier::Path(_) => {
-                bail!("cannot download a package from a local path");
-            }
-        };
-
-        let package = wasmer_registry::query_package_from_registry(
-            api_endpoint.as_str(),
-            full_name,
-            version.as_deref(),
-            token.as_deref(),
-        )
-        .with_context(|| {
-            format!(
+                .with_context(|| {
+                    format!(
                 "could not retrieve package information for package '{}' from registry '{}'",
-                full_name, api_endpoint,
+                full_name, endpoint,
             )
-        })?;
+                })?;
 
-        let download_url = package
-            .pirita_url
-            .context("registry does provide a container download container download URL")?;
+                let download_url = package
+                    .pirita_url
+                    .context("registry does provide a container download container download URL")?;
+
+                (download_url, token)
+            }
+            PackageSource::Ident(PackageIdent::Hash(hash)) => {
+                let endpoint = self.env.registry_endpoint()?;
+                let token = self.env.get_token_opt().map(|x| x.to_string());
+
+                let client = wasmer_api::WasmerClient::new(endpoint, "wasmer-cli")?;
+                let client = if let Some(token) = &token {
+                    client.with_auth_token(token.clone())
+                } else {
+                    client
+                };
+
+                let rt = tokio::runtime::Runtime::new()?;
+                let pkg = rt.block_on(wasmer_api::query::get_package_release(&client, &hash.to_string()))?
+                    .with_context(|| format!("Package with {hash} does not exist in the registry, or is not accessible"))?;
+
+                (pkg.webc_url, token)
```

Comment:
I believe we fixed this... but we didn't @maminrayej can you address this to just get the hash directly from the backend?

Reply:
We should not use a new query to the backend for this, we can just retrieve the hash directly

Reply:
There is no download URL in the frontend/backend to download by hash correctly.

The backend will need to implement this, so it's a followup task.

Reply:
Got it, resolving the question